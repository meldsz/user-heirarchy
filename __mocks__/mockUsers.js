export const mockUsers = {
  secrets: {
    ABCD: 123,
    GHYI: 567,
  },
  users: [
    {
      email: "john.smith@xtreet.tvl",
      firstName: "John",
      id: 123,
      lastName: "Smith",
      password: "asak",
    },
    {
      email: "matthew.bartolacci@btthew.net",
      firstName: "Matthew",
      id: 567,
      lastName: "Bartolacci",
      password: "x6Ldx7Ye",
      photo: "https://test.jpg",
    },
    {
      email: "jane.smith@xtreet.tvl",
      firstName: "Jane",
      id: 777,
      managerId: 567,
      lastName: "Smith",
      password: "uiu",
    },
    {
      email: "jane.doe@xtreet.tvl",
      id: 444,
      password: "iur",
    },
  ],
};
