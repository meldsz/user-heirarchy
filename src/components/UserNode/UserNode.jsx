import { useState } from "react";

export const UserNode = ({ user }) => {
  const [expanded, setExpanded] = useState(false);
  const [isBadImage, setIsBadImage] = useState(false);
  const isManager = user?.children?.length > 0;
  const initials =
    user?.firstName && user?.lastName
      ? user?.firstName[0] + user?.lastName[0]
      : "N/A";

  return (
    <div data-testid="user-node-container">
      <div className="user-container">
        <button
          data-testid="toggle"
          className="toggle"
          onClick={() => setExpanded(!expanded)}
          aria-label={expanded ? "Collapse" : "Expand"}
        >
          {isManager ? "+" : <strong>-</strong>}
        </button>
        {user?.photo && !isBadImage ? (
          <img
            data-testid="user-badge"
            className="user-badge"
            src={user?.photo}
            onError={() => setIsBadImage(true)}
          />
        ) : (
          <div className="user-badge">{initials}</div>
        )}
        <div className="user-details">
          <div>
            {user?.firstName} {user?.lastName}
          </div>
          <div> {user?.email}</div>
        </div>
      </div>

      {isManager && expanded && (
        <div className="user-children">
          {user?.children.map((child) => (
            <UserNode key={child?.id} user={child} />
          ))}
        </div>
      )}
    </div>
  );
};
