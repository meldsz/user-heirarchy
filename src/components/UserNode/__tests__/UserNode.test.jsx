import { fireEvent } from "@testing-library/dom";
import { testRender } from "../../../common/tests/helper";
import { UserNode } from "../UserNode";
import { mockUsers } from "../../../../__mocks__/mockUsers";

describe("<UserNode/>", () => {
  it("should render correctly", () => {
    const { getByTestId, asFragment } = testRender(
      <UserNode user={mockUsers?.users[0]} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("user-node-container")).toBeDefined();
  });
  it("should expand manager correctly", () => {
    const { getByTestId, getByText, getByLabelText } = testRender(
      <UserNode user={mockUsers?.users[0]} />
    );
    expect(getByLabelText("Expand")).toBeDefined();
    fireEvent.click(getByTestId("toggle"));
    expect(getByLabelText("Collapse")).toBeDefined();
  });
  it("should display user image when it is available", () => {
    const { getByTestId, getByRole } = testRender(
      <UserNode user={mockUsers?.users[1]} />
    );
    fireEvent.click(getByTestId("toggle"));
    expect(getByRole("img").src).toBe("https://test.jpg/");
  });
  it("should display user initials when user image is not available", () => {
    const { getByTestId, getByText, queryByTestId } = testRender(
      <UserNode user={mockUsers?.users[0]} />
    );
    fireEvent.click(getByTestId("toggle"));
    expect(getByText("JS")).toBeDefined();
    expect(queryByTestId("user-badge")).toBeNull();
  });
  it("should display user initials when user image has an error", () => {
    const { getByTestId, getByText, getByRole } = testRender(
      <UserNode user={mockUsers?.users[1]} />
    );
    fireEvent.click(getByTestId("toggle"));
    expect(getByRole("img").src).toBe("https://test.jpg/");
    fireEvent.error(getByRole("img"));
    expect(getByText("MB")).toBeDefined();
  });
});
