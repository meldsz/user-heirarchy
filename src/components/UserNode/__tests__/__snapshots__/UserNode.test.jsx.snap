// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<UserNode/> should render correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="user-node-container"
  >
    <div
      class="user-container"
    >
      <button
        aria-label="Expand"
        class="toggle"
        data-testid="toggle"
      >
        <strong>
          -
        </strong>
      </button>
      <div
        class="user-badge"
      >
        JS
      </div>
      <div
        class="user-details"
      >
        <div>
          John Smith
        </div>
        <div>
           john.smith@xtreet.tvl
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
